lib/evaluate/rte.cc

```diff
@@ -0,0 +1,166 @@
+// that are guaranteed to exist from the C++ standard.
+
+#include "rte.h"
+#include <cerrno>
+#include <cfenv>
+
+namespace Fortran::evaluate::rte {
+
+using namespace Fortran::parser::literals;
+// TODO mapping to <cmath> function to be tested.<cmath> func takes
+// real arg for n
+template<typename HostT> static HostT Bessel_jn(std::int64_t n, HostT x) {
+  return std::cyl_bessel_j(static_cast<HostT>(n), x);
+}
+template<typename HostT> static HostT Bessel_yn(std::int64_t n, HostT x) {
+  return std::cyl_neumann(static_cast<HostT>(n), x);
+}
+
+template<typename HostT> void AddLibmRealHostProcedure(HostRte &hostRte) {
+  using F = FuncPointer<HostT, HostT>;
+  using F2 = FuncPointer<HostT, HostT, HostT>;
+  HostRteProcedureSymbol libmSymbols[]{{"acos", F{std::acos}, true},
+      {"acosh", F{std::acosh}, true}, {"asin", F{std::asin}, true},
+      {"asinh", F{std::asinh}, true}, {"atan", F{std::atan}, true},
+      {"atan", F2{std::atan2}, true}, {"atanh", F{std::atanh}, true},
+      {"bessel_jn", &Bessel_jn<HostT>, true},
+      {"bessel_yn", &Bessel_yn<HostT>, true}, {"cos", F{std::cos}, true},
+      {"cosh", F{std::cosh}, true}, {"sin", F{std::sin}, true},
+      {"sinh", F{std::sinh}, true}, {"sqrt", F{std::sqrt}, true},
+      {"tan", F{std::tan}, true}, {"tanh", F{std::tanh}, true}};
+
+  for (auto sym : libmSymbols) {
+    hostRte.AddProcedure(std::move(sym));
+  }
+}
+
+template<typename HostT> void AddLibmComplexHostProcedure(HostRte &hostRte) {
+  using F = FuncPointer<std::complex<HostT>, const std::complex<HostT> &>;
+  HostRteProcedureSymbol libmSymbols[]{{"acos", F{std::acos}, true},
+      {"acosh", F{std::acosh}, true}, {"asin", F{std::asin}, true},
+      {"asinh", F{std::asinh}, true}, {"atan", F{std::atan}, true},
+      {"atanh", F{std::atanh}, true}, {"cos", F{std::cos}, true},
+      {"cosh", F{std::cosh}, true}, {"sin", F{std::sin}, true},
+      {"sinh", F{std::sinh}, true}, {"sqrt", F{std::sqrt}, true},
+      {"tan", F{std::tan}, true}, {"tanh", F{std::tanh}, true}};
+
+  for (auto sym : libmSymbols) {
+    hostRte.AddProcedure(std::move(sym));
+  }
+}
+
+void HostRte::DefaultInit() {
+  AddLibmRealHostProcedure<float>(*this);
+  AddLibmRealHostProcedure<double>(*this);
+  AddLibmRealHostProcedure<long double>(*this);
+
+  AddLibmComplexHostProcedure<float>(*this);
+  AddLibmComplexHostProcedure<double>(*this);
+  AddLibmComplexHostProcedure<long double>(*this);
+}
+
+void HostFloatingPointEnvironment::SetUpHostFloatingPointEnvironment(
+    FoldingContext &context) {
+  errno = 0;
+  if (feholdexcept(&originalFenv_) != 0) {
+    context.messages().Say(
+        "Folding with host runtime: feholdexcept() failed: %s"_en_US,
+        std::strerror(errno));
+    return;
+  }
+  if (fegetenv(&currentFenv_) != 0) {
+    context.messages().Say(
+        "Folding with host runtime: fegetenv() failed: %s"_en_US,
+        std::strerror(errno));
+    return;
+  }
+#if __x86_64__
```

Comment:
Do we really need to use those architecture specific fields of fenv_t directly? _FPU_GETCW()/_FPU_SETCW() pair of macros are defined in <fpu_control.h> which should be available to all glibc-based systems, this could help us avoid use of preprocessor-based checking for architecture (#if __x86_64__) - each place like that makes it less possible to use f18 as a cross-compiler and can make it harder to be accepted by LLVM people.

Reply:
I think the `#ifdef` is unavoidable here because `<fpu_control.h>` is unfortunately only about the X87 FPU on X86_64 and not SSE. `_FPU_GETCW`/`_FPU_SETCW` resolve to assembly instructions `fnldcw`/`fnstcw `on x86_64 which are related to the x87 fpu control word (FPCR).
X86_64 FPCR does not give control over denormal value flushing. What matters for X86_64 is the `mxcsr` register which is related to SSE. It needs to be addressed directly for to control denormal value flushing.

Anyway, thanks for bringing the topic and for pointing to these macros for Aarch64 hosts. I will try to move the `#ifdef` in `host.h` file from this PR that should centralize such host dependent behavior (So far, I think having `#ifdef` in `host.h` is preferable compared to having different files for different hosts.). Once this PR is merged, Aarch64 implementation can be added . Will you want to do it?

Reply:
What is the easiest way to test this piece of code?

Reply:
Good point. I did not add a test for that. It cannot be tested yet with a Fortran program because there is not options to control SubnormalFlushing in f18 tool yet.
It can be tested with a C++ program that tries to fold an expression that requires host runtime and set the context with different subnormal flushing options.
I made a small test that use `log(subnormal-value)` to detect subnormal flushing. If subnormal values are flushed log(0) returns -infinity. If subnormal values are not flushed, it returns a negative finite value.
See commit 35ebaf3.
It passes on X86_64 and once the logic is here for Aarch64 in host.cc, it should pass there also.